datasetinsights/estimators/siam_mask.py

```diff
@@ -0,0 +1,2251 @@
+        )
+        logger.info(
+            "unused checkpoint keys:{}".format(len(unused_pretrained_keys))
+        )
+    logger.info("used keys:{}".format(len(used_pretrained_keys)))
+    assert len(used_pretrained_keys) > 0, "load NONE from pretrained checkpoint"
+    return True
+
+
+def load_pretrain(model, pretrained_path):
+    """
+        Args : Model path
+        Loads a pre-trained resnet model to init the siamese net
+    """
+    logger.info("load pretrained model from {}".format(pretrained_path))
+    if not torch.cuda.is_available():
+        pretrained_dict = torch.load(
+            pretrained_path, map_location=lambda storage, loc: storage
+        )
+    else:
+        device = torch.cuda.current_device()
+        pretrained_dict = torch.load(
+            pretrained_path,
+            map_location=lambda storage, loc: storage.cuda(device),
+        )
+
+    if "state_dict" in pretrained_dict.keys():
+        pretrained_dict = remove_prefix(
+            pretrained_dict["state_dict"], "module."
+        )
+    else:
+        pretrained_dict = remove_prefix(pretrained_dict, "module.")
+
+    try:
+        check_keys(model, pretrained_dict)
+    except AssertionError:
+        logger.info(
+            '[Warning]: using pretrain as features. \
+            Adding "features." as prefix'
+        )
+        new_dict = {}
+        for k, v in pretrained_dict.items():
+            k = "features." + k
+            new_dict[k] = v
+        pretrained_dict = new_dict
+        check_keys(model, pretrained_dict)
+    model.load_state_dict(pretrained_dict, strict=False)
+    return model
+
+
+# Parent class conataining helper functions for
+# freezing weights of the layers at any point of training
+class MultiStageFeature(nn.Module):
+    def __init__(self):
+        super(MultiStageFeature, self).__init__()
+        self.feature_size = -1
+        self.layers = []
+        self.train_num = -1
+        # Will be overriden from child class ideally
+        # This basically says, reading from right to left:
+        # when model is 0.5 trained increase trainable layers to 3
+        # change_point [0, 0.5]
+        # train_nums[1, 3]
+        self.change_point = []
+        self.train_nums = []
+
+    def forward():
+        raise NotImplementedError
+
+    # This ratio is the percentage of training completed
+    # (current_epoch/total_epochs)
+    def unfix(self, ratio=0.0):
+        if self.train_num == -1:
+            self.train_num = 0
+            self.unlock()
+            self.eval()
+        for p, t in reversed(list(zip(self.change_point, self.train_nums))):
+            if ratio >= p:
+                if self.train_num != t:
+                    self.train_num = t
+                    self.unlock()
+                    return True
+                break
+        return False
+
+    # Set which layers will be trainable
+    def train_layers(self):
+        return self.layers[: self.train_num]
+
+    def unlock(self):
```

Comment:
docstring should be clear about `unfix` vs `unlock` as the words are quite similar